dataactvalidator/scripts/loadTas.py

```diff
@@ -74,44 +109,97 @@ def update_tas_lookups(csv_path):
+        fill_in_cols = ['account_title', 'budget_bureau_code', 'budget_bureau_name', 'budget_function_code',
+                        'budget_function_title', 'budget_subfunction_code', 'budget_subfunction_title',
+                        'reporting_agency_aid', 'reporting_agency_name']
+        for _, row in relevant_old_data.iterrows():
+            fill_in_updates = {fill_in_col: row[fill_in_col] for fill_in_col in fill_in_cols}
+            fill_in_updates['updated_at'] = datetime.now(timezone.utc)
+            sess.query(TASLookup).filter_by(account_num=row['account_num']).update(synchronize_session=False,
+                                                                                   values=fill_in_updates)
+        logger.info('%s records filled in', len(relevant_old_data.index))
+    else:
+        # instead of using the pandas to_sql dataframe method like some of the other domain load processes, iterate
+        # through the dataframe rows so we can load using the orm model (note: toyed with the SQLAlchemy bulk load
+        # options but ultimately decided not to go outside the unit of work for the sake of a performance gain)
+        for _, row in old_data.iterrows():
+            sess.query(TASLookup).filter_by(account_num=row['account_num']).update(row, synchronize_session=False)
+
+        for _, row in new_data.iterrows():
+            sess.add(TASLookup(**row))
+        logger.info('%s records in CSV, %s existing', len(data.index), sum(data['existing_id'].notnull()))
 
     sess.commit()
-    logger.info('%s records in CSV, %s existing', len(data.index), sum(data['existing_id'].notnull()))
 
 
-def load_tas(tas_file=None):
-    """Load TAS file into broker database. """
-    # read TAS file to dataframe, to make sure all is well
-    # with the file before firing up a db transaction
+def load_tas(tas_file=None, backfill_historic=False):
+    """ Load TAS file into broker database.
+
+        Args:
+            tas_file: path of the car_tas csv to import
+            backfill_historic: if set to true, this will only update certain columns if budget_function_code is null
+    """
+    # read TAS file to dataframe, to make sure all is well with the file before firing up a db transaction
+    tas_files = []
     if not tas_file:
         if CONFIG_BROKER["use_aws"]:
-            s3connection = boto.s3.connect_to_region(CONFIG_BROKER['aws_region'])
-            s3bucket = s3connection.lookup(CONFIG_BROKER['sf_133_bucket'])
-            tas_file = s3bucket.get_key("cars_tas.csv").generate_url(expires_in=600)
+            # Storing version dictionaries in the list to prevent getting all the links at once and possibly work with
+            # expired AWS links
+            s3connection = boto3.client('s3', region_name=CONFIG_BROKER['aws_region'])
+            # list_object_versions returns the versions in reverse chronological order
+            tas_files = s3connection.list_object_versions(Bucket=CONFIG_BROKER['sf_133_bucket'],
```

Comment:
Do we need to get the versions if we aren't backfilling? Could this part be inside the `if` so we can just grab the latest and not worry about versions if they aren't needed?

Reply:
Understood. Was doing it for code simplicity but I agree it can be taxing in the future.